pkg/git/git_keychain.go

```diff
@@ -4,56 +4,67 @@ import (
 	"sort"
 	"strings"
 
-	git2go "github.com/libgit2/git2go/v33"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	gitssh "github.com/go-git/go-git/v5/plumbing/transport/ssh"
+
 	"github.com/pkg/errors"
 	giturls "github.com/whilp/git-urls"
 	"golang.org/x/crypto/ssh"
 
 	"github.com/pivotal/kpack/pkg/secret"
 )
 
-type Git2GoCredential interface {
-	Cred() (*git2go.Credential, error)
-}
+type CredentialType string
 
-func keychainAsCredentialsCallback(gitKeychain GitKeychain) git2go.CredentialsCallback {
-	return func(url string, usernameFromUrl string, allowedTypes git2go.CredentialType) (*git2go.Credential, error) {
-		cred, err := gitKeychain.Resolve(url, usernameFromUrl, allowedTypes)
-		if err != nil {
-			return nil, err
-		}
-		return cred.Cred()
-	}
+const (
+	CredentialTypeUserpass  CredentialType = "userpass"
+	CredentialTypeSSHKey    CredentialType = "sshkey"
+	CredentialTypeSSHCustom CredentialType = "sshcustom"
+	CredentialTypeDefault   CredentialType = "default"
+	CredentialTypeUsername  CredentialType = "username"
+	CredentialTypeSSHMemory CredentialType = "sshmemory"
+)
+
+type GoGitCredential interface {
+	Cred() (transport.AuthMethod, error)
 }
 
-type GitKeychain interface {
-	Resolve(url string, usernameFromUrl string, allowedTypes git2go.CredentialType) (Git2GoCredential, error)
+type GoGitSshCredential struct {
+	GoGitCredential
+	Username   string
+	Signer     ssh.Signer
+	PrivateKey []byte
 }
 
-type BasicGit2GoAuth struct {
-	Username, Password string
+type GoGitHttpCredential struct {
+	GoGitCredential
+	Username string
+	Password string
 }
 
-func (b BasicGit2GoAuth) Cred() (*git2go.Credential, error) {
-	return git2go.NewCredentialUserpassPlaintext(b.Username, b.Password)
+type GitKeychain interface {
```

Comment:
Remove layer of indirection, look at old implementation here: https://github.com/pivotal/kpack/blob/d4de54b914365f6082bcbe938fa038aab18ed9b9/pkg/git/git_keychain.go#L18